contract/r/gnoswap/pool/upgrade.gno

```diff
@@ -0,0 +1,117 @@
+//
+// The initializer function creates a new instance of the implementation
+// using the provided poolStore interface.
+//
+// Security: Only contracts within the domain path can register initializers.
+// Each package path can only register once to prevent duplicate registrations.
+func RegisterInitializer(cur realm, initializer func(poolStore IPoolStore) IPool) {
+	// Ensure the caller is within the domain path (e.g., pool/v1, pool/v2)
+	if !isContainDomainPath() {
+		panic("RegisterInitializer can only be called in the domain path")
+	}
+
+	previousRealm := runtime.PreviousRealm()
+
+	// Check if this package path has already been registered
+	previousPath := previousRealm.PkgPath()
+	if _, ok := initializers[previousPath]; ok {
+		panic("RegisterInitializer can only be called once")
+	}
+
+	// Register the initializer function for this package path
+	initializers[previousPath] = initializer
+	initialized := poolImpl != nil
+
+	if initialized {
+		// If an implementation is already active, add this version as read-only
+		// This allows the new version to be registered but not immediately active
+		err := kvStore.AddAuthorizedCaller(previousRealm.Address(), store.ReadOnly)
+		if err != nil {
+			panic(err)
+		}
+	} else {
+		// If no implementation is active yet, make this the active implementation
+		poolImpl = initializers[previousPath](NewPoolStore(kvStore))
+
+		// Grant write permissions to the active implementation
+		err := kvStore.AddAuthorizedCaller(previousRealm.Address(), store.Write)
+		if err != nil {
+			panic(err)
+		}
+	}
+}
+
+// UpgradeImpl switches the active pool implementation to a different version.
+// This function allows seamless upgrades from one version to another without
+// data migration or downtime.
+//
+// Security: Only admin addresses can perform upgrades.
+// The new implementation must have been previously registered via RegisterInitializer.
+func UpgradeImpl(cur realm, packagePath string) {
+	// Ensure only admin can perform upgrades
+	caller := runtime.PreviousRealm().Address()
+	access.AssertIsAdmin(caller)
+
+	// Verify the target implementation has been registered
+	if _, ok := initializers[packagePath]; !ok {
+		panic("Initializer not found for package path:" + packagePath)
+	}
+
+	// Switch to the new implementation
+	poolImpl = initializers[packagePath](NewPoolStore(kvStore))
+	packageAddress := chain.PackageAddress(packagePath)
+
+	// Update storage permissions for all registered implementations
+	storeAuthorizedCallers, err := kvStore.GetAuthorizedCallers()
+	if err != nil {
+		panic(err)
+	}
+
+	// Set all implementations to read-only except the new active one
+	for caller, _ := range storeAuthorizedCallers {
+		updatedPermission := store.ReadOnly
+
+		// Only the new active implementation gets write permissions
+		isCurrentPackageAddress := caller == packageAddress
+		if isCurrentPackageAddress {
+			updatedPermission = store.Write
+		}
```

Comment:
I'm not sure if this actually happen, but if a transaction interferes during a permission update, it appears that multiple implementations could temporaily have `Write` permissions. However, this could be prevented by halting the contract and then update.

It would be good to check the halt status ar the proxy layer and verify the halt state in `UpdateImpl` method.